lib/foreman_remote_execution_core/polling_script_runner.rb

```diff
@@ -0,0 +1,138 @@
+module ForemanRemoteExecutionCore
+  class PollingScriptRunner < ScriptRunner
+
+    def initialize(options = {})
+      super(options)
+      @callback_host = options[:callback_host]
+      @task_id = options[:uuid]
+      @step_id = options[:step_id]
+      try_obtain_otp!
+    end
+
+    def refresh_interval
+      ForemanRemoteExecutionCore.settings[:runner_refresh_interval].to_i
+    end
+
+    def prepare_start
+      super
+      basedir         = File.dirname @remote_script
+      @pid_path       = File.join(basedir, 'pid')
+      @retrieval_script ||= File.join(basedir, 'retrieve')
+      prepare_retrieval unless @prepared
+    end
+
+    def control_script
+      close_stdin = '</dev/null'
+      close_fds = close_stdin + ' >/dev/null 2>/dev/null'
+      # pipe the output to tee while capturing the exit code in a file, don't wait for it to finish, output PID of the main command
+      <<-SCRIPT.gsub(/^\s+\| /, '')
+      | sh -c '(#{su_prefix}#{@remote_script} #{close_stdin}; echo $?>#{@exit_code_path}) | /usr/bin/tee #{@output_path} >/dev/null; #{callback_scriptlet}' #{close_fds} &
+      | echo $! > '#{@pid_path}'
+      SCRIPT
+    end
+
+    def trigger(*args)
+      run_sync(*args)
+    end
+
+    def refresh
+      err = output = nil
+      with_retries do
+        _, output, err = run_sync("#{su_prefix} #{@retrieval_script}")
+      end
+      lines = output.lines
+      result = lines.shift.match(/^DONE (\d+)?/)
+      publish_data(lines.join, 'stdout') unless lines.empty?
+      publish_data(err, 'stderr') unless err.empty?
+      if result
+        exitcode = result[1] || 0
+        publish_exit_status(exitcode.to_i)
+      end
+      destroy_session
+    end
+
+    def external_event(event)
+      publish_data(event.data['output'], 'stdout')
+      publish_exit_status(event.data['exit_code'].to_i)
+      super
+    end
+
+    def close
+      super
+      SmartProxyDynflowCore::OtpManager.drop_otp(@uuid, @otp) if @otp
```

Comment:
In the `try_obtain_otp` we are checking if the `SmartProxyDynflowCore` is present. We should be consistent with the expactations, probably extracting this logic to `foreman-tasks-core`